Implementation of RFC-0097: Unbonding Queue for pallet-staking-async

moliholy opened the pull request:
This PR continues the implementation of [rossbulat](https://github.com/paritytech/polkadot-sdk/pull/5715) and [runcomet](https://github.com/paritytech/polkadot-sdk/pull/7398) regarding [RFC-0097](https://polkadot-fellows.github.io/RFCs/approved/0097-unbonding_queue.html) for fast unbonding.

After #8127 being merged, this PR continues the previous work started in #8232, moving the logic from `pallet-statking` to `pallet-staking-async`.

-----

# Description
This PR implements [RFC-0097](https://polkadot-fellows.github.io/RFCs/approved/0097-unbonding_queue.html) and its [modification](https://hackmd.io/@vKfUEAWlRR2Ogaq8nYYknw/SyfioMGWgl), which introduces a variable unbonding time mechanism to improve the security-liquidity trade-off in the staking system.

The unbonding queue allows for faster unbonding of staked tokens while maintaining network security by keeping a minimum slashable share of stake locked for the full unbonding period.

### Modified extrinsics:
- `unbond`: Enhanced to work with the unbonding queue mechanism, calculating variable unbonding times based on stake distribution.
- `withdraw_unbonded`: Modified to handle withdrawals from the unbonding queue with different unlock periods.
- `rebond`: Updated to work with the queue-based unbonding system.

### Storage items:
- `ElectableStashes`: Added the stake backing up a given electable stash account.
- `UnbondingQueueParams`: New storage item containing configuration parameters:
   - `min_slashable_share`: The minimum share of stake of the lowest backed validators that must remain slashable at any point in time.
    - `lowest_ratio`: The proportion of the lowest-backed validator set.
    - `unbond_period_lower_bound`: Minimum unbonding time in eras.
- `ErasLowestRatioTotalStake`: Tracks the lowest stake proportion among validators in each era.
- `TotalUnbondInEra`: Tracks the stake that started unbonding in a given era.
- `Ledger`: Added `previous_unbonded_stake` to track the accumulated stake to be unbonded at the time a given unlock chunk was created.

### View functions:
  - `unbonding_duration`: Used to obtain the list of funds that will be released at a given era.

### Key features:
1. Variable unbonding time: Unbonding time varies based on the amount of stake in the system.
2. Security maintenance: Ensures a sufficient percentage of stake remains slashable at all times.
3. Unbonding queue management: Implements an efficient queue system for managing unbonding requests.
    
### Migration `v18`:
Includes multi-block migration `LazyMigrationV17ToV18` that:
  - Updates `StakingLedger` structure to include the new `previous_unbonded_stake` field in `UnlockChunk`.
  - Migrates `ElectableStashes` from set-based to map-based storage (AccountId -> Balance mapping).
  - Performs era adjustment for existing unlock chunks by subtracting the bonding duration.
  - Uses stepped migration pattern to handle large datasets efficiently across multiple blocks.
  - Includes comprehensive pre/post-upgrade checks for data integrity.

Moliholy commented:
/cmd prdoc

kianenigma commented on substrate/frame/staking-async/src/pallet/impls.rs:
```diff
@@ -70,6 +70,11 @@ use sp_runtime::TryRuntimeError;
 /// invalid (for any reason) the iteration continues. With this constant, we iterate at most 2 * n
 /// times and then give up.
 const NPOS_MAX_ITERATIONS_COEFFICIENT: u32 = 2;
+const PICO_ERA: u128 = 1_000_000_000_000;
```
Isn't this an overkill? An era is 24h, would't a percent of it be enough accuracy to slice it into smaller pieces?

Moliholy replied:
The idea is to make an attack to push the back of the unbonding queue economically unfeasible. We performed so tests in this regard to calculate how much it'd cost to push it one unit. Here's the script:

```javascript
const {ApiPromise, WsProvider} = require('@polkadot/api')

const getValidatorsWithStake = async () => {
    const wsProvider = new WsProvider('wss://rpc.polkadot.io')
    const api = await ApiPromise.create({provider: wsProvider})
    const validators = await api.query.session.validators()
    const activeEra = await api.query.staking.activeEra()
    const eraIndex = activeEra.unwrap().index

    console.log(`Current era: ${eraIndex}`)
    console.log(`Total validators: ${validators.length}`)

    // Get stake information for each validator
    const validatorsWithStake = (await Promise.all(
        validators.map(async validatorId => {
            // Query the exposure for each validator in the current era
            const exposure = (await api.query.staking.erasStakersOverview(eraIndex, validatorId)).toJSON()

            return {
                validator: validatorId.toString(),
                totalStake: BigInt(exposure.total),
                ownStake: BigInt(exposure.own),
                nominatorsCount: exposure.nominatorCount,
            }
        })
    ))
        .sort((a, b) => Number(a.totalStake - b.totalStake))
    const lowestThirdStakeValidators = validatorsWithStake.slice(0, Math.ceil(validators.length / 3))
    const totalLowestThirdStake = lowestThirdStakeValidators.reduce((acc, {totalStake}) => acc + totalStake, BigInt(0)) / BigInt(10000000000)
    const maxUnstake = totalLowestThirdStake / 2n
    const maxUnbondZeroDeltaEra = maxUnstake / 28n
    const maxUnbondZeroDeltaPicoEra = Number(maxUnbondZeroDeltaEra) / Math.pow(10, 12)
    const maxUnbondZeroDeltaBlock = maxUnstake / 403200n

    console.log(`Lowest third validators total stake: ${totalLowestThirdStake} DOT`)
    console.log(`max_unstake: ${maxUnstake} DOT`)
    console.log(`max_unbond_zero_delta_era: ${maxUnbondZeroDeltaEra} DOT`)
    console.log(`max_unbond_zero_delta_pico_era: ${maxUnbondZeroDeltaPicoEra} DOT`)
    console.log(`max_unbond_zero_delta_block: ${maxUnbondZeroDeltaBlock} DOT`)
}

getValidatorsWithStake()
    .catch(console.error)
    .finally(() => process.exit())
```

And these are the results:

```
Current era: 1802
Total validators: 600
Lowest third validators total stake: 248302884 DOT
max_unstake: 124151442 DOT
max_unbond_zero_delta_era: 4433980 DOT
max_unbond_zero_delta_pico_era: 0.00000443398 DOT
max_unbond_zero_delta_block: 307 DOT
```

So, this results should that it'd be necessary to spend 0.00000443398 DOT + transaction fee to increase by MORE than one pico-era the back of the unbonding queue.

With an amount of minimum 10**-10 DOT + transaction fees, since we round up, an attacker would increase by one pico-era. So, the idea is that an attacker would spend a relatively huge amount of funds to artificially push the queue.

If we'd be using whole eras, an tiny amount could potentially push a whole era, which is far from ideal. We could indeed use nano-eras or micro-eras as well, but that'd also increase by orders of magnitud the delay to the back of the unbonding queue, since the smallest unit would be much bigger now. In my opinion, using pico-eras provides enough robustness for different scenarios and works as a long-term solution.

But not an strong opinion at all, and I'll happily change it if needed be.

kianenigma replied:
I think we should explore the implications of the transaction that would allow anyone to `poke` their unbonding delta. This might alleviate the risks associated with this, no?

Moliholy replied:
> to poke their unbonding delta

It's not "their" unbonding delta, but everyone else's that unbonds after the operation is performed. According to the calculations, and assuming a transaction fee on AH of 0.0011087708 DOT, it'd take 0.0011087709 * 10**12 = 1.1 billion DOT to push one era up with this attack, which is economically unfeasible, since each transaction would increment the back of the unbonding queue by an extremely low value.

To calculate the cost using nano-eras, or pico-eras, simply divide by 1_000 or 1_000_000 respectively.

kianenigma commented on substrate/frame/staking-async/src/pallet/mod.rs:
```diff
@@ -868,6 +886,9 @@ pub mod pallet {
 			if let Some(x) = self.max_nominator_count {
 				MaxNominatorsCount::<T>::put(x);
 			}
+			if let Some(x) = self.unbonding_queue_config {
```
```suggestion
			UnbondingQueueParams::<T>::put(self.unbonding_queue_config);
```

Moliholy replied:
Changed in [aa3d40c](https://github.com/paritytech/polkadot-sdk/pull/8298/commits/aa3d40c2cb608696e9669c2a5f1bea022fd4848d).

kianenigma commented on substrate/frame/staking-async/src/pallet/impls.rs:
```diff
@@ -856,6 +861,129 @@ impl<T: Config> Pallet<T> {
+		if let Some(params) = UnbondingQueueParams::<T>::get() {
+			// Determine the total stake from the lowest portion of validators and persist for the
+			// era.
+			let mut exposures: Vec<_> =
```
This is not exposure, this is `exposure_total`, or better said, `total_stake`. I think using `exposure` is misleading as it might imply the type is `Exposure`.

Moliholy replied:
Changed in [aa3d40c](https://github.com/paritytech/polkadot-sdk/pull/8298/commits/aa3d40c2cb608696e9669c2a5f1bea022fd4848d).

kianenigma commented on substrate/frame/staking-async/src/pallet/impls.rs:
```diff
@@ -856,6 +861,129 @@ impl<T: Config> Pallet<T> {
+	pub fn get_min_lowest_stake() -> BalanceOf<T> {
+		// Find the minimum total stake of the lowest portion validators over the configured number
+		// of eras.
+		EraLowestRatioTotalStake::<T>::get().into_iter().min().unwrap_or(Zero::zero())
```
```suggestion
		EraLowestRatioTotalStake::<T>::get().into_iter().min().unwrap_or(Bounded::max_value())
```

This is a better default, as it would lead to all unbonding times being the maximum, while this queue is not filled. 

Even better is to use an `Option` here -- if the past 28 eras are not filled, it should return `None`, which means we should not process any ubonding requests "faster".

Moliholy replied:
> This is a better default, as it would lead to all unbonding times being the maximum, while this queue is not filled.

Quite the opposite. This will increase `max_unstake` to its maximum, which will make any arbitrary amount to be unbonded in the minimum time.

The more stake in validators, the more a user can unbond without penalty, as the system has enough stake to slash.

kianenigma commented on substrate/frame/staking-async/src/tests/unbonding_queue.rs:
```diff
@@ -0,0 +1,603 @@
+				assert_ok!(Staking::bond_extra(RuntimeOrigin::signed(21), 100));
+				assert_ok!(Staking::bond_extra(RuntimeOrigin::signed(31), 100));
+				assert_eq!(Staking::get_min_lowest_stake(), 500);
+				Session::roll_until_active_era(i + 1);
```
You are already in era 2, this will be a noop in the first round.

Moliholy replied:
Completely refactored in [aa3d40c](https://github.com/paritytech/polkadot-sdk/pull/8298/commits/aa3d40c2cb608696e9669c2a5f1bea022fd4848d).

kianenigma commented on substrate/frame/staking-async/src/lib.rs:
```diff
@@ -153,6 +153,35 @@ pub struct ActiveEraInfo {
+	/// The minimum unbonding time, in eras, for an active stake.
+	pub unbond_period_lower_bound: EraIndex,
+	/// The pico-era when all the existing unbonders have unbonded.
+	pub back_of_unbonding_queue: u128,
```
This gets updated all the time, I don't think it should be part of the condigs.

Moliholy replied:
Yes, right. I simply didn't want to add yet another extra storage item, as the pallet already has a few. I'll change it.

kianenigma commented on substrate/frame/staking-async/src/tests/unbonding_queue.rs:
```diff
@@ -0,0 +1,603 @@
+use sp_runtime::{traits::Zero, Perbill};
+
+#[test]
+fn get_min_lowest_stake_works() {
```
I have questions about the methodology of how this test is written. 

as far as I can see, this test wants to assert that the lowest 1/3 total stake is correctly tracked. 

So what you would need to test is: 

1. First, as discussed elsewhere, if we don't have enough previous data, this should return `None`. Note that in genesis, the `ExtBuilder` will roll only 1 era. So first we want to assert that it does that.
2. Then, you want to roll forward until enough eras such that we gathered enough data that it return `Some(_)`. 
3. But, in the above, all of the previous total stakes are the same, so getting all of the previous values are the same
4. Then, you'd want to tweak at least one validator to have less/more total stake, then roll one more active era, and then assert that the queue of total backings has indeed now changed.

Moliholy replied:
> First, as discussed elsewhere, if we don't have enough previous data, this should return None.

As already explained, returning zero is both the correct value (there is a total of zero accumulated stake over the last `BondDuration` eras for the lowest third validators) and it also yields the correct number of unbonding eras,  since `max_unstake = 0`. `max_unstake` represents the amount of stake that yields the maximum number of eras when unbonding, so a value of zero means any arbitrary amount will yield the maximum, which is the correct behaviour.

As for the rest, I completely refactored the test and made it much more explicit in [aa3d40c](https://github.com/paritytech/polkadot-sdk/pull/8298/commits/aa3d40c2cb608696e9669c2a5f1bea022fd4848d).

kianenigma commented on substrate/frame/staking-async/src/tests/unbonding_queue.rs:
```diff
@@ -0,0 +1,603 @@
+				})
+			));
+
+			assert_eq!(Staking::get_min_lowest_stake(), 0);
```
Also I don't like naming things exactly as per the RFC, as they might not translate well in the code. 

For example, the name `get_min_lowest_stake` doesn't imply what this value exactly is, unless if the RFC is read.

Moliholy replied:
So... I agree with you: it's not intuitive enough. But the method represents the minimum stake in the last 28 eras that covers the lowest third validators with the lowest stake. How can you possible find a name to express all that and not to occupy a whole line?

I hoped it'd suffice with the method's description, which I'll happily change if needed be for a more descriptive version. If not... does any function name ring a bell?

kianenigma commented on substrate/frame/staking-async/src/tests/unbonding_queue.rs:
```diff
@@ -0,0 +1,603 @@
+		]);
+
+		// Trigger new era to calculate the lowest proportion.
+		assert_ok!(EraElectionPlanner::<T>::do_elect_paged_inner(
```
This is shortcutting some code -- why not create the validator/nominator, then `Session::roll_until_active_era(next)`? this is even less verbose, and it is the realistic code path.

kianenigma commented on substrate/frame/staking-async/src/tests/unbonding_queue.rs:
```diff
@@ -0,0 +1,603 @@
+		assert_ok!(EraElectionPlanner::<T>::do_elect_paged_inner(
+			exposures.clone().try_into().unwrap()
+		));
+		Session::roll_until_active_era(5);
```
Yeah this is strange, the above line should not be needed.

Moliholy replied:
Why? The test fails if it gets removed.

Moliholy replied:
Test refactored in [4d21ca1](https://github.com/paritytech/polkadot-sdk/pull/8298/commits/4d21ca1e36ea10cff917a9b7d97c3e02c82b636c).

kianenigma commented on substrate/frame/staking-async/src/tests/unbonding_queue.rs:
```diff
@@ -0,0 +1,603 @@
+
+		Session::roll_until_active_era(2);
+		Session::roll_until_active_era(3);
+		Session::roll_until_active_era(4);
```
You can just do `roll_until_active_era(4)` and remove 2 and 3.

Moliholy replied:
Done in [1866f07](https://github.com/paritytech/polkadot-sdk/pull/8298/commits/1866f076aa40cc6001a19fdad2b83be4a74ce359).

kianenigma commented on substrate/frame/staking/src/pallet/impls.rs:
```diff
@@ -1990,6 +1990,10 @@ impl<T: Config> StakingInterface for Pallet<T> {
 		T::BondingDuration::get()
 	}
 
+	fn max_bonding_duration(_value: Self::Balance) -> EraIndex {
```
should be removed.

Moliholy replied:
Removed in [0f7d125](https://github.com/paritytech/polkadot-sdk/pull/8298/commits/0f7d12504cd6620004e2107ba8d08ab67e3f3861).

kianenigma commented on substrate/primitives/staking/src/lib.rs:
```diff
@@ -197,9 +197,14 @@ pub trait StakingInterface {
 
+	/// Number of eras that `value` staked funds must remain bonded for.
+	/// TODO: this should be renamed to `bonding_duration` or alike.
+	fn max_bonding_duration(value: Self::Balance) -> EraIndex;
```
What we do need instead is a set of `#[pallet::view_function]` that allow wallets to query what is the current unbonding time for a given amount and similar scenarios.

Moliholy replied:
> Why is this added to begin with?

Because now the existing trait is not accurate. Now we can query two things:

- The unbonding duration given an amount.
- The maximum unbonding duration the system allows.

The second one is already present in the API, but its name is no longer correct, and should in any case be refactored as `max_unbonding_duration()`, but that's a breaking change.

As for the first one, it definitely makes sense to include it if we include the other one.

As for the view function, fine for me (nice addition, btw! 🎊 ), but that has nothing to do with the `StakingInterface` trait, which might be used in other parts of the system internally.

Moliholy replied:
View function added in [7cdf802](https://github.com/paritytech/polkadot-sdk/pull/8298/commits/7cdf802c7dbe6a267b02ec79001e1aab13bf547d).

kianenigma commented on substrate/frame/staking-async/src/tests/unbonding_queue.rs:
```diff
@@ -0,0 +1,603 @@
+		// Must be the maximum because the minimum lowest stake is zero, which defaults to
+		// the maximum unbonding period.
+		assert_eq!(EraLowestRatioTotalStake::<Test>::get().into_inner(), vec![]);
+		assert_eq!(Staking::get_unbonding_delta(1, config), 3000000000000);
```
or your `normalize_era(3)`.

Moliholy replied:
Done in [1866f07](https://github.com/paritytech/polkadot-sdk/pull/8298/commits/1866f076aa40cc6001a19fdad2b83be4a74ce359).

kianenigma commented on substrate/frame/staking-async/src/tests/unbonding_queue.rs:
```diff
@@ -0,0 +1,603 @@
+#[test]
+fn get_unbond_eras_delta_works() {
+	ExtBuilder::default().build_and_execute(|| {
+		let config = UnbondingQueueConfig {
```
Again, I urgue you to: 

1. Remove `back_of_unbonding_queue` from the `UnbondingQueueConfig`, it is not a configuration but rather a state variable. 
2. `UnbondingQueueConfig` should be more clear in what values should be set such that this feature is a noop
3. A helper function `ExtBuilder` should enable it for said tests.

Moliholy replied:
First part refactored in [1866f07](https://github.com/paritytech/polkadot-sdk/pull/8298/commits/1866f076aa40cc6001a19fdad2b83be4a74ce359).

Moliholy replied:
Second part done in [88449ef](https://github.com/paritytech/polkadot-sdk/pull/8298/commits/88449ef4718d8c8867bf7b0501ed528e52fdb797).

Moliholy replied:
Actually I took the liberty of doing it that way in [3d05285](https://github.com/paritytech/polkadot-sdk/pull/8298/commits/3d052850f1266521b012d5f8e30f6c04bdf0d81d), as it makes the tests cleaner.

Moliholy commented on substrate/frame/staking-async/src/tests/unbonding_queue.rs:
```diff
@@ -0,0 +1,603 @@
+		.set_status(31, StakerStatus::<AccountId>::Validator)
+		.validator_count(3)
+		.build_and_execute(|| {
+			assert_ok!(Staking::set_staking_configs(
```
Parameters have to all be specified (`ConfigOp::Noop`), even if they are not changed. I don't come up with a solution that doesn't require using macros nor making the logic more complicated, so if you have any suggestion on how to make it cleaner I'd be happy to implement it.

Moliholy commented:
@kianenigma the pallet's design has been updated as per the received feedback and it is ready to be reviewed again.

Ank4n commented on substrate/frame/staking-async/src/session_rotation.rs:
```diff
@@ -751,6 +751,13 @@ impl<T: Config> Rotator<T> {
+		if starting_era >= diff {
+			let target_era = starting_era - diff;
+			TotalUnbondInEra::<T>::remove(target_era);
+			EraLowestRatioTotalStake::<T>::remove(target_era);
```
`prune_era` might be a better place for this. This function only cleans up data that we maintain per era cycle, and not historical.

Moliholy commented on substrate/frame/staking-async/src/pallet/mod.rs:
```diff
@@ -1391,20 +1417,35 @@ pub mod pallet {
-				if let Some(chunk) = ledger.unlocking.last_mut().filter(|chunk| chunk.era == era) {
+				let current_era = CurrentEra::<T>::get().unwrap_or(0);
+				let previous_unbonded_stake =
+					TotalUnbondInEra::<T>::get(current_era).unwrap_or_default();
```
Yes, that's the accumulated amount of all unbonds in a given era, so when a new era starts its default value is zero.

Moliholy commented on substrate/frame/staking-async/src/pallet/mod.rs:
```diff
@@ -1391,20 +1417,35 @@ pub mod pallet {
 				};
+				let new_total_unbond_in_era = {
+					let val = previous_unbonded_stake.defensive_saturating_add(value);
+					if val.is_zero() {
```
Hmmm yes, it cannot be zero, as there is a check a few lines above. Nice catch!

Moliholy commented on substrate/frame/staking-async/src/ledger.rs:
```diff
@@ -345,16 +350,46 @@ impl<T: Config> StakingLedger<T> {
-					true
-				} else {
-					total = total.saturating_sub(chunk.value);
+				if current_era >= chunk.era.defensive_saturating_add(T::BondingDuration::get()) {
```
Yes, exactly. I thought to rename the parameter, but that'd make things more complicated regarding the migration and on a second thought I was fine with just documenting it.

Moliholy commented on substrate/frame/staking-async/src/pallet/mod.rs:
```diff
@@ -1391,20 +1417,35 @@ pub mod pallet {
-					.unwrap_or(0)
-					.defensive_saturating_add(T::BondingDuration::get());
-				if let Some(chunk) = ledger.unlocking.last_mut().filter(|chunk| chunk.era == era) {
+				let current_era = CurrentEra::<T>::get().unwrap_or(0);
```
Previously `CurrentEra` was used here, so that makes me wonder whether what was before was a mistake or not. Would prefer to have more eyes on this to determine if we are to change it now.

Ank4n replied:
This is def needs to be fixed but outside your PR scope so you can ignore it. We are working on fixing this use of `Current Era` in the code in other places as well, and depending on which PR is merged first, the other will have to update accordingly.

Moliholy replied:
I guess you mean `Rotator::<T>::planned_era()`. Done in [96d4226](https://github.com/paritytech/polkadot-sdk/pull/8298/commits/96d4226fdd0dea7ded102147819d05f7c24fd1ed), but I saw many other places through the pallet where this should be changed at some point in a separate PR.

Moliholy commented on substrate/frame/staking-async/src/pallet/mod.rs:
```diff
@@ -799,8 +801,29 @@ pub mod pallet {
 	#[pallet::storage]
-	pub type ElectableStashes<T: Config> =
-		StorageValue<_, BoundedBTreeSet<T::AccountId, T::MaxValidatorSet>, ValueQuery>;
+	pub type ElectableStashes<T: Config> = StorageValue<
```
I see drawbacks with that approach:

1) Storing the electable stash's corresponding support makes sense. It is the justification that stash is actually included in that storage item for.
2) If storing in `EraLowestRatioTotalStake`, we'd have duplicated addresses with both storage items. Not optimal for storage efficiency, plus there could be conciliation issues due to the duplicated information. Not to mention we'd have to wait a full 28-era period for it to be cleaned up, while with the current solution they are cleaned up upon a new era is reached.
3) The bounded map is not sorted by balance, but by account. That implies a O(n) operation per removal/insertion or a O(n log n) to sort the map and then remove/insert more efficiently. Either way degrading current performance.

Moliholy commented on substrate/frame/staking-async/src/pallet/mod.rs:
```diff
@@ -799,8 +801,29 @@ pub mod pallet {
+	/// This is used to determine the maximum amount of stake that can be unbonded for a period
+	/// potentially lower than upper-bound eras.
+	#[pallet::storage]
+	pub type EraLowestRatioTotalStake<T: Config> =
```
Definitely not a fan of these getter/setters, but here ([3e57c94](https://github.com/paritytech/polkadot-sdk/pull/8298/commits/3e57c94ef32ce50cde66d9cab2c4f29e60153bd6)) it is.

Ank4n commented on substrate/frame/staking-async/src/pallet/impls.rs:
```diff
@@ -880,6 +885,92 @@ impl<T: Config> Pallet<T> {
+				if final_era <= current_era {
+					free.saturating_accrue(chunk.value);
+				} else {
+					result.insert(final_era, chunk);
```
Also, its possible for multiple chunks to resolve to same `final_era`, which is not desirable.

Moliholy replied:
> Can you clarify?

We are not resetting `unbonding_start_era` here. That's a value that, once set upon creation of the chunk, never changes. `final_era` is the dynamically-calculated era when funds in that chunk can be released, given the current system conditions. If not included in the curated unlocking queue, it means funds in that chunk can be released immediately.

> Also, its possible for multiple chunks to resolve to same final_era, which is not desirable.

If several chunks can be released in past eras, it does not matter at all in which particular ones they were actually available to be released. What matters is that they could eventually have been released in the past but were not, and hence they can be released right now. All those funds get accumulated in the `free` variable.

In case of yet-to-be-released chunks, I didn't manage to reproduce a case where several chunks (created in different eras) can be released in the same future era. Anyhow, I added [0f1ae18](https://github.com/paritytech/polkadot-sdk/pull/8298/commits/0f1ae1876ed889ca50eaed7e86c3e014cbe3b221) to allow a list of chunks to be released per era.

Ank4n commented on substrate/frame/staking-async/src/pallet/impls.rs:
```diff
@@ -880,6 +885,92 @@ impl<T: Config> Pallet<T> {
+						(Perbill::from_percent(100) - min_slashable_share) * lowest_stake;
+					if total_unbond >= threshold {
+						final_era = final_era
+							.max(chunk.era.defensive_saturating_add(T::BondingDuration::get()));
```
In all cases the second item (`chunk.era + upper_bound`) will be bigger right?

Moliholy replied:
Indeed, but that's actually a bug I just corrected. This is the correct version:

```rust
final_era = final_era.max(era.defensive_saturating_add(T::BondingDuration::get()));
```

Ank4n commented on substrate/frame/staking-async/src/pallet/impls.rs:
```diff
@@ -880,6 +885,92 @@ impl<T: Config> Pallet<T> {
+					} else {
+						era_total_amount
+					};
+					total_unbond.saturating_accrue(unbond);
```
`total_unbond` is always same as `unbond` since you are reinitialising `total_unbond` in every iteration.

Moliholy replied:
Indeed. As you mentioned, `total_unbond` is meant to accumulate over several iterations. That was fixed in previous commits.

Ank4n commented on substrate/frame/staking-async/src/pallet/impls.rs:
```diff
@@ -880,6 +885,92 @@ impl<T: Config> Pallet<T> {
+			} else {
+				let mut final_era = chunk.era.defensive_saturating_add(min_unlock_era);
+				for era in (target_era..=chunk.era).rev() {
+					let mut total_unbond = BalanceOf::<T>::zero();
```
Total unbond needs to accumulate across eras.

Moliholy replied:
Indeed, it must be accumulated. Done in [a8217ba](https://github.com/paritytech/polkadot-sdk/pull/8298/commits/a8217bacd49cec6792aa760a8f135b4f1930883e).

Ank4n reviewed (changes requested):
The current withdraw logic seems to deviate quite a bit from the original spec, without clear explanation of why or how. It would really help to align the code more closely with the spec—or at least document the intentional deviations.

I’d also suggest simplifying the flow where possible and improving variable names/comments in critical paths like consolidate_unlocked. Right now, parts of the logic are hard to follow and could be prone to subtle bugs.

There are also a few small issues (e.g. misleading accumulation logic) that reduce confidence in both correctness and test coverage.

Personally, I find the spec quite clear—looping over each unlock chunk and testing eligibility directly—and would prefer an implementation that mirrors that logic more closely.

Moliholy commented on substrate/frame/staking-async/src/benchmarking.rs:
```diff
@@ -128,6 +128,29 @@ pub(crate) fn create_validator_with_nominators<T: Config>(
+	)
+	.expect("failed to set staking configs");
+
+	for _ in 0..T::BondingDuration::get() {
+		Eras::<T>::set_lowest_stake(0, 1000u32.into());
```
Actually this is not needed anymore, as it is part of the initial RFC, which got changed in the second iteration.

Moliholy commented on substrate/frame/staking-async/src/benchmarking.rs:
```diff
@@ -1185,6 +1222,60 @@ mod benchmarks {
+		let mut unlocking = vec![];
+		for _ in 0..c {
+			unlocking
+				.push(crate::migrations::v18::v17::UnlockChunk { value: 100_u32.into(), era: 10 });
```
Hmmmm that's true, although it does not affect the migration itself. Changed in [20af53e](https://github.com/paritytech/polkadot-sdk/pull/8298/commits/20af53e31c9d8cf2aa7041859c7cdd285c00433a).

Moliholy commented on substrate/frame/staking-async/src/ledger.rs:
```diff
@@ -54,9 +54,12 @@ pub struct UnlockChunk<Balance: HasCompact + MaxEncodedLen> {
 	#[codec(compact)]
 	pub value: Balance,
-	/// Era number at which point it'll be unlocked.
+	/// Era number when the chunk was created.
```
Improved documentation in [d0597b0](https://github.com/paritytech/polkadot-sdk/pull/8298/commits/d0597b0c9bf9f689e7d6b180cc534a1bce3bf54e).

Moliholy commented on substrate/frame/staking-async/src/pallet/impls.rs:
```diff
@@ -880,6 +885,92 @@ impl<T: Config> Pallet<T> {
+			Ok(l) => l,
+			Err(_) => return ((current_era, Zero::zero()), BTreeMap::new()),
+		};
+		let target_era = current_era.saturating_add(1).saturating_sub(T::BondingDuration::get());
```
Variable renamed in [a8217ba](https://github.com/paritytech/polkadot-sdk/pull/8298/commits/a8217bacd49cec6792aa760a8f135b4f1930883e).

Moliholy commented:
@Ank4n thanks a lot for the feedback. Ready for another round.

kianenigma commented on substrate/frame/staking-async/ahm-test/src/ah/mock.rs:
```diff
@@ -491,6 +491,11 @@ impl ExtBuilder {
+			unbonding_queue_config: Some(pallet_staking_async::UnbondingQueueConfig {
+				min_slashable_share: Perbill::from_percent(50),
+				lowest_ratio: Perbill::from_percent(34),
+				unbond_period_lower_bound: 2,
```
I would do both: add upper_bound here, and keep the old one, but marked as #[deprecated].

Moliholy replied:
There is a corner case for that, which is in `[OffenceQueueEras](https://github.com/paritytech/polkadot-sdk/blob/master/substrate/frame/staking-async/src/pallet/mod.rs#L736)`.  It is certainly useful to have the maximum bonding duration as a configuration item for two reasons:

1) No storage read needed.
2) Can be used to determine maximum sizes.

One alternative in this case would be to create a separate configuration item to specify the maximum size of the `OffenceQueueEras`, but nothing guarantees that the maximum bonding duration changes in the future, as it could be dynamically adjusted without the need of a runtime upgrade.

Ank4n replied:
>  renaming it is also a breaking change for clients currently using it.

IMO, a breaking change isn’t always bad. Since the meaning of this config has changed significantly, it’s better to rename it and explicitly make it a breaking change rather than keep the old name.

kianenigma commented on substrate/frame/staking-async/ahm-test/Cargo.toml:
```diff
@@ -66,3 +64,17 @@ try-runtime = [
 	"pallet-staking-async-ah-client/try-runtime",
 	"pallet-timestamp/try-runtime",
 ]
+runtime-benchmarks = [
```
sure thet are needed? this pallet has not benchmark code.

kianenigma commented on substrate/frame/staking-async/src/pallet/impls.rs:
```diff
@@ -880,6 +885,98 @@ impl<T: Config> Pallet<T> {
+		if let Some(params) = UnbondingQueueParams::<T>::get() {
+			// Determine the total stake from the lowest portion of validators and persist for the
+			// era.
+			let mut total_stake: Vec<_> =
```
```suggestion
			let mut validator_total_stakes: Vec<_> =
```

This can be mistaken with the `.sum` of all of them, which we already track in `ErasTotalStake`.

Moliholy replied:
Done in [2079d89](https://github.com/paritytech/polkadot-sdk/pull/8298/commits/2079d899504b69781d0374a7836dd2c24c68dacd).

kianenigma commented on substrate/frame/staking-async/src/pallet/impls.rs:
```diff
@@ -880,6 +885,98 @@ impl<T: Config> Pallet<T> {
+	pub(crate) fn curate_unlocking_chunks(
+		stash: T::AccountId,
+	) -> ((EraIndex, BalanceOf<T>), BTreeMap<EraIndex, Vec<UnlockChunk<BalanceOf<T>>>>) {
+		let current_era = CurrentEra::<T>::get().unwrap_or(Zero::zero());
```
As noted elsewhere, plasea read this from `struct Eras` for consitency.

Moliholy replied:
Refactored to use `Rotator::<T>::planned_era()`.

Moliholy commented:
@kianenigma @Ank4n after https://github.com/paritytech/polkadot-sdk/pull/9079 was merged (while this PR has been awaiting reviews), we are now facing a severe issue, which is essentially that offences expect an absolute unbonding era. However, after the changes introduced in this PR, the unbonding era is dynamic and subject to the system's global stake, which causes conflicts with the implementation when withdrawing. This will require to redesign and reimplement offences processing logic, and that is a substantial effort. We request Parity's help on this considering the changes have been introduced while this PR was pending.

Moliholy commented on substrate/frame/staking-async/src/pallet/mod.rs:
```diff
@@ -801,8 +802,29 @@ pub mod pallet {
+
+	/// The amount of stake that started unbonding in a given era.
+	#[pallet::storage]
+	pub type TotalUnbondInEra<T: Config> =
```
Done in [f2972c4](https://github.com/paritytech/polkadot-sdk/pull/8298/commits/f2972c459e40cd150028aa02231e484c21ac79e5).

Ank4n commented on substrate/frame/staking-async/src/migrations/v18.rs:
```diff
@@ -0,0 +1,408 @@
+	weights::WeightMeter,
+};
+
+pub(crate) mod v17 {
```
Nit: I think I mentioned this earlier but can’t find it now - putting v17 under v18 is confusing. I would rather split them so v17 is a sibling module (e.g., add v17.rs) and keep old structs under migrations::v17, not v18::v17.

Ank4n replied:
Found here: https://github.com/paritytech/polkadot-sdk/pull/8298#discussion_r2172641973

I am not saying we should not have the old types under v17. What I find odd is having v17 nested inside v18. As in your [example](https://github.com/paritytech/polkadot-sdk/blob/master/substrate/frame/identity/src/migration.rs#L66), v1 and v2 are siblings.

Moliholy replied:
The actual multi-block migration [example](https://github.com/paritytech/polkadot-sdk/blob/master/substrate/frame/examples/multi-block-migrations/src/migrations/v1/mod.rs#L50) uses that exact approach...

Also, `v17` is not accessible outside the pallet, and hence not exposed.

Ank4n commented on substrate/frame/staking-async/src/migrations/v18.rs:
```diff
@@ -0,0 +1,408 @@
+	Noop,
+}
+
+pub struct LazyMigrationV17ToV18<T: Config>(PhantomData<T>);
```
Looks solid - thanks. Do you have an estimate of how many blocks this migration will take on Polkadot AH, based on current ledger size?

Ank4n commented on substrate/frame/staking-async/src/pallet/impls.rs:
```diff
@@ -934,6 +927,100 @@ impl<T: Config> Pallet<T> {
+			let total_stake = validator_total_stakes.into_iter().sum();
+
+			// Store the total stake of the lowest portion validators for the planned era.
+			Eras::<T>::set_lowest_stake(era, total_stake)
```
Please confirm there are adequate tests (using realistic Polkadot/Kusama-scale validator sets) and try-state checks to ensure lowest_stake is calculated correctly.

Ank4n commented on substrate/frame/staking-async/src/pallet/impls.rs:
```diff
@@ -934,6 +927,100 @@ impl<T: Config> Pallet<T> {
+		let current_era = Rotator::<T>::planned_era();
+		let ledger = match Self::ledger(Stash(stash)) {
+			Ok(l) => l,
+			Err(_) => return ((current_era, Zero::zero()), BTreeMap::new()),
```
This should never happen right? You could add a defensive check here.

Ank4n commented on substrate/frame/staking-async/runtime-api/src/lib.rs:
```diff
@@ -35,5 +38,8 @@ sp_api::decl_runtime_apis! {
 		fn pending_rewards(era: sp_staking::EraIndex, account: AccountId) -> bool;
+
+		/// Returns a list of (era, amount) that indices at which era unbonded funds will be unlocked.
+		fn unbonding_duration(account: AccountId) -> Vec<(sp_staking::EraIndex, Balance)>;
```
Suggestion: We should expose a runtime api that that estimates `if I unbond now, when do my funds unlock?` so frontends can show the current ETA. Wdyt?

Ank4n commented on substrate/frame/staking-async/src/benchmarking.rs:
```diff
@@ -1268,6 +1301,62 @@ mod benchmarks {
+		for era in 0..c {
+			unlocking.push(crate::migrations::v18::v17::UnlockChunk {
+				value: 100_u32.into(),
+				era: 10 + era,
```
wouldn't this fail against runtimes with bonding duration greater than 10?

Moliholy replied:
No, this is just blindly migrating what is in the storage. Any arbitrary number would work for this. Anyway, it could be initially set to `MaxUnbondingDuration`. Done in [468edb6](https://github.com/paritytech/polkadot-sdk/pull/8298/commits/468edb6f1fd94e5ea70c71a376bbef62f70312de).

Ank4n commented on substrate/frame/staking-async/src/pallet/mod.rs:
```diff
@@ -2162,6 +2200,7 @@ pub mod pallet {
 			config_op_exp!(ChillThreshold<T>, chill_threshold);
 			config_op_exp!(MinCommission<T>, min_commission);
 			config_op_exp!(MaxStakedRewards<T>, max_staked_rewards);
+			config_op_exp!(UnbondingQueueParams<T>, unbonding_queue_params);
```
Should we have some integrity checks here? Such as lower bound should never exceed `UnbondingDuration`.

Moliholy commented on substrate/primitives/staking/src/lib.rs:
```diff
@@ -197,7 +197,7 @@ pub trait StakingInterface {
 	fn stash_by_ctrl(controller: &Self::AccountId) -> Result<Self::AccountId, DispatchError>;
 
-	/// Number of eras that staked funds must remain bonded for.
+	/// Maximum number of eras that staked funds must remain bonded for.
```
Done in [6c98c6b](https://github.com/paritytech/polkadot-sdk/pull/8298/commits/6c98c6b023d4a269980265ee001efa8d2202380f).

Moliholy commented on substrate/frame/staking-async/src/pallet/mod.rs:
```diff
@@ -2162,6 +2203,14 @@ pub mod pallet {
+			if let ConfigOp::Set(params) = unbonding_queue_params {
+				ensure!(
+					params.unbond_period_lower_bound <= T::MaxUnbondingDuration::get(),
+					Error::<T>::InvalidConfiguration
```
Sounds very reasonable. Changed in [6a30cb4](https://github.com/paritytech/polkadot-sdk/pull/8298/commits/6a30cb4971c85e2716cbf4963a0613e136351f1e).

Moliholy commented on substrate/frame/staking-async/src/pallet/impls.rs:
```diff
@@ -211,27 +224,14 @@ impl<T: Config> Pallet<T> {
 			.and_then(|eras| eras.first())
 			.copied()
 			// if nothing in queue, use the active era.
 			.unwrap_or(active_era)
```
Yes, you are right. I simply changed logical comparison, including equality. But `calculate_earliest_withdrawal_era` was indeed not returning by itself the correct value. I just fixed it in [e69da63](https://github.com/paritytech/polkadot-sdk/pull/8298/commits/e69da63902f3205723df4cf22710199440d54b2e).

Moliholy commented:
@kianenigma @Ank4n we have wrapped up the work here from BlockDeep side and have addressed the review feedback provided so far. We are handing it over to you and the code owners for any future changes and merging the PR when possible. Thank you.

kianenigma commented on substrate/frame/staking-async/src/session_rotation.rs:
```diff
@@ -808,6 +837,11 @@ impl<T: Config> Rotator<T> {
 		EraElectionPlanner::<T>::cleanup();
 
+		if let Some(target_era) = starting_era.checked_sub(T::MaxUnbondingDuration::get()) {
+			Eras::<T>::remove_total_unbond_for_era(target_era);
```
I don't think we can ever remove this information -- as long as there is anyone that hasn't called withdraw for their old funds, we need to keep them.

sigurpol commented:
As mentioned [here](https://forum.polkadot.network/t/proposal-dynamic-allocation-pool-dap/15878/20?u=sigurpol) and considering the follow-up [post](https://forum.polkadot.network/t/the-roadmap-for-the-dynamic-allocation-pool-dap/16511) and the related [WFC](https://polkadot.subsquare.io/referenda/1827), we believe we should drop RFC-97 and do not continue with this PR. The reason is that nominators will benefit for fast unbonding:
> Nominator’s stake can unbond as soon as they are not backing any active validators anymore. Depending on when the unbonding extrinsic was cast, this takes between a minimum of 24 and a maximum of 48 hours.

Fast unbonding is arguably less critical for validators. In the future, we might want to consider enabling fast unbonding for a validator that has never entered the active set.

A related PR is currently under review: https://github.com/paritytech/polkadot-sdk/pull/10502 and aimed to be merged ASAP and deployed together with the other relevant changes for DAP / phase 1 by March.

Polkadot-Forum commented:
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/the-roadmap-for-the-dynamic-allocation-pool-dap/16511/8
